src/handlers/expiry/providers/TestDataProvider.ts

```diff
@@ -0,0 +1,90 @@
+import { HTTPError } from "../../../models/HTTPError";
+import {
+  TEST_STATUS,
+  MESSAGES,
+  TEST_CODES_FOR_CALCULATING_EXPIRY
+} from "../../../assets/Enums";
+
+import { ITestResult } from "../../../models/ITestResult";
+import { TestResultsDAO } from "../../../models/TestResultsDAO";
+import { DateProvider } from "./DateProvider";
+import { ITestDataProvider } from "./ITestDataProvider";
+import { Service } from "../../../models/injector/ServiceDecorator";
+
+// TODO: Add unit tests data provider
+// 1- missing attributes scenario
+// 2- postivie scenario
+
+@Service()
+export class TestDataProvider implements ITestDataProvider {
+    public testResultsDAO: TestResultsDAO | undefined;
+
+  // getTestResult
+// Do we use existing functionality or write new function
+  public async getTestHistory(systemNumber: string): Promise<ITestResult[]> {
+    // testService object
+    const fromDateTime = new Date(1970, 1, 1);
+    const toDateTime = new Date();
+    let result: ITestResult[] = [];
+    try {
+      this.testResultsDAO = this.testResultsDAO as TestResultsDAO;
+      const data = await this.testResultsDAO.getBySystemNumber(systemNumber);
+      if (data && data.Count) {
+        result = data.Items as ITestResult[];
+        result = result.filter(
+          (testResult: {
+            testStartTimestamp: string | number | Date;
+            testEndTimestamp: string | number | Date;
+          }) => {
+            return (
+              moment(testResult.testStartTimestamp).isAfter(fromDateTime) &&
+              moment(testResult.testEndTimestamp).isBefore(toDateTime)
+            );
+          }
+        );
+        result = result.filter(
+          (test) => test.testStatus === TEST_STATUS.SUBMITTED
+        );
+      }
+      return result;
+    } catch (error) {
+      if (!(error instanceof HTTPError)) {
+        console.log(error);
+        error = new HTTPError(500, MESSAGES.INTERNAL_SERVER_ERROR);
+      }
+      throw error;
+    }
+  }
+
+  public async getMostRecentExpiryDate(systemNumber: string): Promise<Date> {
+    let maxDate = DateProvider.getEpoc();
+    const testResults = await this.getTestHistory(systemNumber);
+    if (testResults && testResults.length) {
+      const filteredTestTypeDates: any[] = [];
+      testResults.forEach((testResult) => {
+        testResult.testTypes.forEach((testType) => {
+          if (
+            testType.testCode &&
+            TestDataProvider.isValidTestCodeForExpiryCalculation(
+              testType.testCode.toUpperCase()
+            ) &&
+            testType.testExpiryDate
+              &&
+            DateProvider.isValidDate(testType.testExpiryDate)
+          ) {
+            filteredTestTypeDates.push(moment(testType.testExpiryDate));
+          }
+        });
+      });
+      if (filteredTestTypeDates && filteredTestTypeDates.length > 0) {
+        maxDate = moment.max(filteredTestTypeDates).toDate();
+      }
+    }
+    return maxDate;
```

Comment:
Is it worth adding a few more when  `testExpiryDate` are valid so we can see the output of the function in the unit test with the maxDate?

Reply:
Added one scenario with all valid dates and added one scenario with an invalid date.